src/core/include/mp-units/numbers.h

```diff
@@ -0,0 +1,299 @@
+concept number_line =  // clang-format off
+  ordered_number<T> &&
+  requires(T& v) {
+    number_one_v<number_difference_t<T>>;
+    { ++v } -> std::same_as<T&>;
+    { --v } -> std::same_as<T&>;
+    { v-- } -> std::same_as<T>;
+    { v-- } -> std::same_as<T>;
+  };  // clang-format on
+
+namespace detail {
+
+template<typename T, typename U>
+concept addition_with =  // clang-format off
+  number<T> &&
+  number<U> &&
+  requires( const T & c,  const U & d) {
+    { c + d } -> common_number_with<T>;
+    { d + c } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_addition_with =  // clang-format off
+  addition_with<T,U> &&
+  requires(T& l,  const U & d) {
+    { l += d } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept subtraction_with =  // clang-format off
+  addition_with<T,U> &&
+  requires( const T & c,  const U & d) {
+    { c - d } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_subtraction_with =  // clang-format off
+  subtraction_with<T,U> &&
+  compound_addition_with<T,U> &&
+  requires(T& l,  const U & d) {
+    { l -= d } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U, typename V>
+concept multiplication_with =  // clang-format off
+  number<T> &&
+  number<U> &&
+  number<V> &&
+  requires( const T & c,  const U & u) {
+    { (c * u) } -> common_number_with<V>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_multiplication_with =  // clang-format off
+  multiplication_with<T,U,T> &&
+  requires(T& l,  const U & u) {
+    { l *= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept division_with =  // clang-format off
+  multiplication_with<T,U,T> &&
+  multiplication_with<U,T,T> &&
+  requires( const T & c,  const U & u) {
+    { c / u } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_division_with =  // clang-format off
+  division_with<T,U> &&
+  compound_multiplication_with<T,U> &&
+  requires(T& l,  const U & u) {
+    { l /= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+}  // namespace detail
+
+template<typename T, typename U>
+concept modulo_with =  // clang-format off
+  number<T> &&
+  number<U> &&
+  requires( const T & c,  const U & u) {
+    { (c % u) } -> common_number_with<T>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept compound_modulo_with =  // clang-format off
+  modulo_with<T,U> &&
+  requires(T& l,  const U & u) {
+    { l %= u } -> std::same_as<T&>;
+  };  // clang-format on
+
+template<typename T, typename U>
+concept modulus_for = modulo_with<U, T>;
+template<typename T, typename U>
+concept compound_modulus_for = compound_modulo_with<U, T>;
+
+template<typename T>
+concept negative =  // clang-format off
+  detail::compound_addition_with<T,T> &&
```

Comment:
It requires that the set has addition (<https://www.electropedia.org/iev/iev.nsf/display?openform&ievref=102-01-14>).
That's done through `compound_addition_with` because I haven't had an use case where that didn't work.

Reply:
`safe::var` can support addition just fine, but compound addition is the problem. Can we separate out the compound requirements?

Reply:
Of course.

Reply:
Here's a short synopsis for why this breaks safe arithmetic:

```c++
using namespace safe::literals;

// declare variable `x` to be between -10 and 10 inclusive, guaranteed at compile time:
safe::ival_s32<-10, 10> x{};

// adding `1` to `x` results in a different type:
safe::ival_s32<-9, 11> y = x + 1_s32;

// compound addition cannot work because y cannot be assigned to x:
static_assert(false == std::is_assignable_v<decltype(x), decltype(y)>);
```

https://github.com/intel/safe-arithmetic

Reply:
This one requires compound addition with the same type.
Is that supported?